279 voting power is not synced properly on start of epoch

Vitomir2 opened the pull request:
# Description

- delete `epoch.Validators.Copy()` on creating `NewValidatorSet` in order to avoid redundant operations;
- optimize if conditions for the end and start of epoch;
- get the validators from the last epoch;
- update the implementation for the generation of the `syncValidatorsData` input to properly extract the removed validators, too;
- add a test for the function which includes added, updated and removed validators in the validators delta;
- new testing `createTestExtraForSyncValidatorsData` function;

# Changes include

- [x] Bugfix (non-breaking change that solves an issue)
- [ ] Hotfix (change that solves an urgent issue, and requires immediate attention)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (change that is not backwards-compatible and/or changes current functionality)

# Checklist

- [ ] I have assigned this PR to myself
- [x] I have added at least 1 reviewer
- [x] I have added the relevant labels
- [ ] I have updated the official documentation
- [x] I have added sufficient documentation in code

## Testing

- [x] I have tested this code with the official test suite
- [x] I have tested this code manually

### Manual tests

I have configured the blockchain locally and have tested if the syncValidatorsData's input is generated correctly and it seems all work fine. Additionally, added some tests to ensure that the extract function works properly.

Vitomir2 committed:
update sync validators data functionality

use epoch.Validators.Copy() in order to ensure that the validators are not passed by reference and updated in some of the next functions;
optimize if conditions for the end and start of epoch;
get the validators from the last epoch;
update the implementation for the generation of the syncValidatorsData input to properly extract the removed validators, too;

Vitomir2 committed:
add tests

Vitomir2 committed:
delete unused vars

R-Santev commented on consensus/polybft/consensus_runtime_test.go:
```diff
@@ -782,6 +783,149 @@ func TestConsensusRuntime_calculateStateTxsInput_SecondEpoch(t *testing.T) {
+
+	validatorsDelta := parentIbftExtraData.Validators
+
+	expectedUpdatedValidatorsPower := extractUpdatedValidatorsVotingPower(validatorsDelta, valSetAccounts, accSet)
```
instead of using additional function, can't you hardcode the expectedUpdatedValidatorsPower

R-Santev commented on consensus/polybft/consensus_runtime_test.go:
```diff
@@ -782,6 +783,149 @@ func TestConsensusRuntime_calculateStateTxsInput_SecondEpoch(t *testing.T) {
 	polybftBackendMock.AssertExpectations(t)
 }
 
+func TestConsensusRuntime_generateSyncValidatorsDataTxInputRemoved(t *testing.T) {
```
maybe better having a combined test where there are added updated and deleted validators

Vitomir2 commented on consensus/polybft/validator/validator_metadata.go:
```diff
@@ -367,27 +367,39 @@ func (as AccountSet) ApplyDelta(validatorsDelta *ValidatorSetDelta) (AccountSet,
-			updatedValidatorsPower = append(updatedValidatorsPower, formatValidatorPower(validator))
+	// Check if we have added or updated validators first before looping through the current validators
+	// We need to loop through the current validators because they have the updated voting power
+	if validatorsDelta.Added.Len() > 0 || validatorsDelta.Updated.Len() > 0 {
```
Yeah, it will be better to do it like that. For some reason, I was confused that I am missing the `votingPower` in the validatorSetDelta.

Vitomir2 committed:
updates

delete the .Copy() as discussed with Roskata;
update the ExtractUpdatedValidatorsVotingPower logic to loop through the added, updated and then check removed validators;
avoid setting the VotingPower to 0 and add it as a parameter of the formatValidatorPower function;
create new test function createTestExtraForSyncValidatorsData and use it for the extraData;
make one test that includes added, updated and removed validators;
set concrete params instead of mock.Anything in the GetValidators mock;

Vitomir2 committed:
delete redundant test function

Vitomir2 committed:
add nolint
